contrib/gazebo/gazebo_sawyer_env.py

```diff
@@ -0,0 +1,499 @@
+import threading
+import time
+from functools import wraps
+
+from intera_core_msgs.msg import JointLimits
+
+from gazebo_msgs.msg import (
+    ModelStates, )
+
+import std_srvs.srv
+
+from geometry_msgs.msg import (Pose, Point)
+
+
+class ControlModeUnknownErr(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+def goal_distance(goal_a, goal_b):
+    """
+    :param goal_a:
+    :param goal_b:
+    :return: distance between goal_a and goal_b
+    """
+    assert goal_a.shape == goal_b.shape
+    return np.linalg.norm(goal_a - goal_b, axis=-1)
+
+
+def rate_limited(max_per_second: int):
+    """Rate-limits the decorated function locally, for one process."""
+    lock = threading.Lock()
+    min_interval = 1.0 / max_per_second
+
+    def decorate(func):
+        last_time_called = time.perf_counter()
+
+        @wraps(func)
+        def rate_limited_function(*args, **kwargs):
+            lock.acquire()
+            nonlocal last_time_called
+            try:
+                elapsed = time.perf_counter() - last_time_called
+                left_to_wait = min_interval - elapsed
+                if left_to_wait > 0:
+                    time.sleep(left_to_wait)
+
+                return func(*args, **kwargs)
+            finally:
+                last_time_called = time.perf_counter()
+                lock.release()
+
+        return rate_limited_function
+
+    return decorate
+
+
+class GazeboSawyerEnv(GazeboEnv, Serializable):
+    """
+    Gazebo Sawyer Env
+    """
+
+    def __init__(self,
+                 has_object,
+                 distance_threshold,
+                 initial_qpos,
+                 reward_type,
+                 target_in_the_air,
+                 limb='right',
+                 end_point="right_gripper_tip",
+                 target_range=0.15,
+                 obj_range=0.15,
+                 control_mode='position'):
+        """
+        Input
+        -----
+        has_object: if there is object in this experiment
+        distance_threshold: this is used for sparse reward computation
+        initial_qpos: dict, joint's positions and object's positions
+        reward_type: 'sparse' of 'dense'
+        target_in_the_ari: set to see if the target can be in the air
+        limb: the name of the limb
+        end_point: the name of the end_point
+        target_range: used to sample the new target
+        obj_range: used to sample the new position for object
+        control_mode: have 3 different control modes: 'position', 'velocity', 'effort'
+        """
+        # run the superclass initialize function first
+        Serializable.quick_init(self, locals())
+
+        rospy.init_node("GazeboSawyerPosControlEnv", anonymous=True)
+        # Verify robot is enabled
+        print("Getting robot state...")
+        self._rs = intera_interface.RobotEnable(intera_interface.CHECK_VERSION)
+        self._init_state = self._rs.state().enabled
+        print("Enabling robot...")
```

Comment:
please don't change hardware state in constructors, or do any other long-running RPC operations. you can do this in reset() instead.